File: python/Q787_cheapest_flights_within_k_stops.py
#
# @lc app=leetcode id=787 lang=python3
#
# [787] Cheapest Flights Within K Stops
#


# @lc code=start
class Solution:
    def findCheapestPrice(
        self, n: int, flights: list[list[int]], src: int, dst: int, k: int
    ) -> int:
        prices: list[int] = [10**5 for _ in range(n)]
        prices[src] = 0

        for _ in range(k + 1):
            new_prices: list[int] = prices.copy()

            for source, destination, price in flights:
                if prices[source] == 10**5:
                    continue

                if (
                    new_prices[destination] == 10**5
                    or new_prices[destination] > prices[source] + price
                ):
                    new_prices[destination] = prices[source] + price

            prices = new_prices

        return prices[dst] if prices[dst] != 10**5 else -1


# @lc code=end

File: rust/src/leetcode/Q787_cheapest_flights_within_k_stops.rs
/*
 * @lc app=leetcode id=787 lang=rust
 *
 * [787] Cheapest Flights Within K Stops
 */

// @lc code=start
impl Solution {
    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        let mut distances: Vec<i32> = vec![i32::MAX; n as usize];
        distances[src as usize] = 0;

        for _ in 0..k + 1 {
            let mut new_distances: Vec<i32> = distances.clone();
            let mut updated: bool = false;

            for flight in &flights {
                let from: usize = flight[0] as usize;
                let to: usize = flight[1] as usize;
                let price: i32 = flight[2];

                if distances[from] != i32::MAX && distances[from] + price < new_distances[to] {
                    new_distances[to] = distances[from] + price;
                    updated = true;
                }
            }

            distances = new_distances;

            if !updated {
                break;
            }
        }

        if distances[dst as usize] == i32::MAX {
            -1
        } else {
            distances[dst as usize]
        }
    }
}
// @lc code=end
